src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
+            head = newNode;
+            tail = newNode;
+        } else { // Якщо список не порожній
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndexForAdd(index); // Метод перевірки допустимого індексу
+
+        Node<T> newNode = new Node<>(value);
+
+        if (index == 0) { // Додаємо на початок списку
+            if (head == null) { // Якщо список порожній
+                head = newNode;
+                tail = newNode;
+            } else {
+                newNode.next = head;
+                head.prev = newNode;
+                head = newNode;
+            }
+        } else if (index == size) { // Додаємо в кінець списку
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else { // Додаємо в середину списку
+            Node<T> current = getNodeByIndex(index); // Отримуємо вузол на вказаному індексі
+            Node<T> previous = current.prev;
+
+            newNode.next = current;
+            newNode.prev = previous;
+
+            if (previous != null) {
+                previous.next = newNode;
+            }
+            current.prev = newNode;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return; // Якщо список порожній або null, нічого не додаємо
+        }
+
+        for (T element : list) {
+            add(element); // Додаємо кожен елемент у кінець поточного списку
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index); // Перевірка валідності індексу
+
+        Node<T> current = head; // Починаємо з голови списку
+        for (int i = 0; i < index; i++) {
+            current = current.next; // Переходимо до наступного вузла
+        }
+
+        return current.value; // Повертаємо значення вузла за індексом
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index); // Перевіряємо, чи індекс валідний
+
+        Node<T> current = head; // Починаємо з голови списку
+        for (int i = 0; i < index; i++) {
+            current = current.next; // Переходимо до потрібного вузла
+        }
+
+        T oldValue = current.value; // Зберігаємо попереднє значення
+        current.value = value; // Замінюємо значення на нове
+        return oldValue; // Повертаємо попереднє значення
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);// Перевірка валідності індексу
+
+        if (index == 0) {
+            // Якщо видаляємо перший елемент
+            T removedValue = head.value; // Зберігаємо значення вузла
+            head = head.next; // Змінюємо посилання на наступний вузол
```

Comment:
When removing the first element, if the list becomes empty (i.e., `head` becomes `null`), you should also set `tail` to `null` to maintain consistency.